source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -413,6 +419,48 @@ DnsCacheImpl::PrimaryHostInfo::~PrimaryHostInfo() {
   parent_.stats_.num_hosts_.dec();
 }
 
+void DnsCacheImpl::addCacheEntry(const std::string& host,
+                                 const Network::Address::InstanceConstSharedPtr& address) {
+  if (!key_value_store_) {
+    return;
+  }
+  const std::string value = absl::StrCat(address->asString());
+  key_value_store_->addOrUpdate(host, value);
+}
+
+void DnsCacheImpl::removeCacheEntry(const std::string& host) {
+  if (!key_value_store_) {
+    return;
+  }
+  key_value_store_->remove(host);
+}
+
+void DnsCacheImpl::loadCacheEntries(
+    const envoy::extensions::common::dynamic_forward_proxy::v3::DnsCacheConfig& config) {
+  if (!config.has_persistent_cache_config()) {
+    return;
+  }
+  auto& factory =
+      Config::Utility::getAndCheckFactory<KeyValueStoreFactory>(config.persistent_cache_config());
+  key_value_store_ = factory.createStore(config.persistent_cache_config(), validation_visitor_,
+                                         main_thread_dispatcher_, file_system_);
+  KeyValueStore::ConstIterateCb load = [this](const std::string& key, const std::string& value) {
+    auto address = Network::Utility::parseInternetAddressAndPortNoThrow(value);
+    if (address == nullptr) {
+      return KeyValueStore::Iterate::Break;
+    }
+    stats_.cache_load_.inc();
+    std::list<Network::DnsResponse> response;
+    response.emplace_back(Network::DnsResponse(address, std::chrono::seconds(0) /* ttl */));
```

Comment:
By passing in `seconds(0)` does this mean the cache entries are already expired?

Reply:
AFIK ttl in this struct is not used :-(
DNS TTL is filled in, and ignored in favor of the "how long do we persist cache entries" hard coded config.  IMO DNS should persist a max of server-returned TTL and configured "longest time I'm willing to wait" but given it's ignored and a I have a TODO to handle lifetimes I left this for later.

Reply:
*blink* Well that's really something! :) For Envoy as a proxy, I can imagine this is no problem. I won't if we'll end up with any wrinkles in the context of Envoy as a client if we're caching DNS names longer than the advertised TTL?

But back to this PR, perhaps we'll end up wanting to persist the cache expiration time to disk so that when we reload them we can evict any entries that have already expired? (Maybe something to do when you resolve the todo below?)